### ISSUE TITLE
Memory leaks when calling eth.contract.functions

### ISSUE THREAD

**Issue by JPFrancoia** [ethereum/web3.py#2296]
* Version: 5.25.0
* Python: 3.10
* OS: linux

## What is wrong?

Repeatedly calling `eth.contract.functions` causes memory leaks. I found one issue mentioning this problem here: https://github.com/ethereum/web3.py/issues/1126, however I'm not sure the problem was really addressed.

### Snippet to reproduce the issue

Please note that I'm creating the contract only once. The memory leak happens just by calling a function.

```python
import json
from web3 import Web3
from memory_profiler import profile


@profile
def fetch(contract):
    return contract.functions.getReserves().call()


addr = "0xa37cd29A87975f44b83F06F9BA4D51879a99d378"

with open("abi.json") as f_in:
    abi = json.load(f_in)

client = Web3(Web3.HTTPProvider("https://api.avax.network/ext/bc/C/rpc"))
pair_contract = client.eth.contract(address=client.toChecksumAddress(addr), abi=abi)

while True:
    data = fetch(pair_contract)
    print(data)
```

### Contract ABI

For the snippet above to work, you'll need to create a `abi.json` file containing:

```json
[{"inputs":[],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"spender","type":"address"},{"indexed":false,"internalType":"uint256","name":"value","type":"uint256"}],"name":"Approval","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"sender","type":"address"},{"indexed":false,"internalType":"uint256","name":"amount0","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"amount1","type":"uint256"},{"indexed":true,"internalType":"address","name":"to","type":"address"}],"name":"Burn","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"sender","type":"address"},{"indexed":false,"internalType":"uint256","name":"amount0","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"amount1","type":"uint256"}],"name":"Mint","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"sender","type":"address"},{"indexed":false,"internalType":"uint256","name":"amount0In","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"amount1In","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"amount0Out","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"amount1Out","type":"uint256"},{"indexed":true,"internalType":"address","name":"to","type":"address"}],"name":"Swap","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"internalType":"uint112","name":"reserve0","type":"uint112"},{"indexed":false,"internalType":"uint112","name":"reserve1","type":"uint112"}],"name":"Sync","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"from","type":"address"},{"indexed":true,"internalType":"address","name":"to","type":"address"},{"indexed":false,"internalType":"uint256","name":"value","type":"uint256"}],"name":"Transfer","type":"event"},{"constant":true,"inputs":[],"name":"DOMAIN_SEPARATOR","outputs":[{"internalType":"bytes32","name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"MINIMUM_LIQUIDITY","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"PERMIT_TYPEHASH","outputs":[{"internalType":"bytes32","name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"internalType":"address","name":"","type":"address"},{"internalType":"address","name":"","type":"address"}],"name":"allowance","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"spender","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"}],"name":"approve","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"balanceOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"to","type":"address"}],"name":"burn","outputs":[{"internalType":"uint256","name":"amount0","type":"uint256"},{"internalType":"uint256","name":"amount1","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"decimals","outputs":[{"internalType":"uint8","name":"","type":"uint8"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"factory","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"getReserves","outputs":[{"internalType":"uint112","name":"_reserve0","type":"uint112"},{"internalType":"uint112","name":"_reserve1","type":"uint112"},{"internalType":"uint32","name":"_blockTimestampLast","type":"uint32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"_token0","type":"address"},{"internalType":"address","name":"_token1","type":"address"}],"name":"initialize","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"kLast","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"to","type":"address"}],"name":"mint","outputs":[{"internalType":"uint256","name":"liquidity","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"name","outputs":[{"internalType":"string","name":"","type":"string"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"nonces","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"owner","type":"address"},{"internalType":"address","name":"spender","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"},{"internalType":"uint256","name":"deadline","type":"uint256"},{"internalType":"uint8","name":"v","type":"uint8"},{"internalType":"bytes32","name":"r","type":"bytes32"},{"internalType":"bytes32","name":"s","type":"bytes32"}],"name":"permit","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"price0CumulativeLast","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"price1CumulativeLast","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"to","type":"address"}],"name":"skim","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"uint256","name":"amount0Out","type":"uint256"},{"internalType":"uint256","name":"amount1Out","type":"uint256"},{"internalType":"address","name":"to","type":"address"},{"internalType":"bytes","name":"data","type":"bytes"}],"name":"swap","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"symbol","outputs":[{"internalType":"string","name":"","type":"string"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[],"name":"sync","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"token0","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"token1","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"}],"name":"transfer","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"value","type":"uint256"}],"name":"transferFrom","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"}]
```

### Memory profiler data

![leak](https://user-images.githubusercontent.com/2399090/148702390-f0bcae6a-8ec0-4a59-97e6-fd4a7a80f0b4.jpg)

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     61.7 MiB     61.7 MiB           1   @profile
     7                                         def fetch(contract):
     8     61.7 MiB      0.0 MiB           1       return contract.functions.getReserves().call()


[4464847917954919967, 5601168068181174333, 1637405112]

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     61.7 MiB     61.7 MiB           1   @profile
     7                                         def fetch(contract):
     8     62.0 MiB      0.3 MiB           1       return contract.functions.getReserves().call()
```

From time to time, the call to `contract.functions.getReserves().call()` consumes an additional 0.3 MiB of memory. That doesn't seem like a lot, but it adds up pretty quickly, especially if you have several contracts running in parallel. AFAIK, it affects functions like `functions.getAmountsOut` and `getReserves`, but I suspect it might affect other functions.
The memory usage grows linearly with the number of calls.

## How can it be fixed?

I don't know. Looking at the data, it looks like an internal data structure of the contract might be resizing, because the increase in memory usage is jumpy.

Would you have any idea?

**Comment by JPFrancoia** [comment]
UPDATE: I think it comes from this line:

https://github.com/ethereum/web3.py/blob/6a90a26ea12e5a789834c9cd6a7ae6d302648f88/web3/contract.py#L1501

**Comment by JPFrancoia** [comment]
Any lead? I've been staring at the code for 3 days I haven't made any progress. I loose the trail when an object called `Method`  is called (I think this object is called...).

**Comment by kclowes** [comment]
Thanks for the report! All of us maintainers are neck deep in other things right now, but will put this in our queue. If you end up continuing to look into it, any updates are helpful!

**Comment by JPFrancoia** [comment]
Will do. Will also be happy to test anything if your team ends up finding the problem before me. Thanks again

**Comment by v1nvn** [comment]
I am also facing a similar issue. 
@JPFrancoia Were you able to get to the bottom of this memory leak?

**Comment by krivars** [comment]
I'm also seeing this issue.

**Comment by fselmo** [comment]
I cannot reproduce this. I am using a similar contract in my snippet for mainnet Ethereum with a `getReserves()` implementation. Even running with a low number of iterations, a plateau pattern already emerges. However, running with 300 iterations, just to be sure, I get the following...

```py
UNI_V2_FACTORY_ADDRESS = Web3.to_checksum_address(
    "0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"
)
DAI_ADDRESS = "0x6B175474E89094C44Da98b954EedeAC495271d0F"
WETH_ADDRESS = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"

w3 = Web3(Web3.HTTPProvider(HTTP_LOCAL))
uniswap_factory_contract = w3.eth.contract(
    address=UNI_V2_FACTORY_ADDRESS, abi=_get_contract_abi(UNI_V2_FACTORY_ADDRESS)
)


dai_weth_address = uniswap_factory_contract.functions.getPair(
    DAI_ADDRESS, WETH_ADDRESS
).call()
dai_weth_contract = w3.eth.contract(
    address=dai_weth_address, abi=_get_contract_abi(dai_weth_address)
)


def run_iterations(num_iterations=300):
    for i in range(num_iterations):
        print(f"--- Iteration {i + 1} ---")
        r = dai_weth_contract.functions.getReserves().call()
        print(r)


if __name__ == "__main__":
    run_iterations()
```

Running the above script with `mprof run path/to/script` and subsequent plotting with `mprof plot`, this is my result:

<img width="1108" alt="Screenshot 2024-10-08 at 15 25 20" src="https://github.com/user-attachments/assets/48578097-2daa-4c29-a031-a3376ecbd3da">

I'm not sure if this used to be an issue and is no longer an issue (thanks for the patience), or if it has anything to do with your environment and setup, but I cannot reproduce this. I am going to close this issue for now but if anyone points to any evidence of a memory leak we would love to know and will be happy to re-open this or a new issue and track it.

Best of luck.
